### ISSUE TITLE
Asynchronous exceptions from IStreamServerEndpoint.listen are swallowed by strports.service

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8710]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#8710|
|Type|defect|
|Created|2016-07-29 18:49:22Z|
|Branch|https://github.com/twisted/twisted/tree/8710-listen-async-exceptions|

`strports.service` sets `svc._raiseSynchronously` to get a synchronous exception in the case where listening fails synchronously, even though the interface is asynchronous.

However, if the failure is really asynchronous, then by the time the errback runs (`handleIt` in `StreamServerEndpointService.privilegedStartService`), it is too late for appending to `raisedNow` to have any effect, so the failure is essentially silently thrown away.

I think the correct fix is to have `privilegedStartService` reset `_raiseSynchronously` back to `False` before it returns.

(As an aside, I have no idea why this variable is initialized to `None` instead of `False`, especially as it is documented as being of type `bool`.)

<details><summary>Searchable metadata</summary>

```
trac-id__8710 8710
type__defect defect
reporter__mithrandi mithrandi
priority__normal normal
milestone__None None
branch__8710_listen_async_exceptions 8710-listen-async-exceptions
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1469818162335626 1469818162335626
changetime__1470706873955916 1470706873955916
version__None None
owner__mithrandi mithrandi

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

Thanks for the changes!

My understanding is that `self._raiseSynchronously` is a flag which is not supposed to be changed at runtime.

It used as a shortcut to have StreamServerEndpointServiceSync and StreamServerEndpointServiceASync

In t.a.strport.service() we are already touching the private `_raiseSynchronously` attribute and set it to True .... and later after the service is started this value is set to False.

I don't understand what is going on there :)

I feel that this fix is just adding more tech-debt to this functionality

Since I don't understand these changes, I am leaving the ticket in the review queue.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @mithrandi|
|-|-|

I'm approving the change to merge (I don't think it's actually making the situation _worse_), but I think we might want to do a follow-up, since adi's confusion is understandable.  There should probably be a private API that, rather than setting some random flag, is actually the compatibility-constructor called by strports.service.

Also, I think this fix is an improvement to the usual experience, but it's not perfect.  If one of these services is stopped and then started it's going to behave inconsistently now.

In any case, take this feedback for some future tickets but land this one when the tests pass.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|GitHub set status to `closed`|
|-|-|

In changeset 692026cd58a5158997c2aba37f4bd746cc158354
```
#!CommitTicketReference repository="" revision="692026cd58a5158997c2aba37f4bd746cc158354"
Merge pull request #445 from twisted/8710-listen-async-exceptions

Author: mithrandi
Reviewer: glyph
Fixes: #8710 

Stop swallowing asynchronous exceptions in `strports.service` / `twistd`.
```
